### PR TITLE
Docs: Add do's and don'ts to block design documentation

### PR DESCRIPTION
**Pull request by jasmussen**
This PR adds a bunch of examples of what you should do, and what you _shouldn't_ do, when designing a block. It leverages existing principles, but expands with screenshots.

Screenshots:

screenshot 2018-10-26 at 11 09 22
screenshot 2018-10-26 at 11 09 28
screenshot 2018-10-26 at 11 09 52
screenshot 2018-10-26 at 11 09 58
screenshot 2018-10-26 at 11 10 04
screenshot 2018-10-26 at 11 10 11

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
😍 Excited to review this one!

**Comment by kristastevens** [comment]
Hello @jasmussen, this is in great shape. I have but one small suggestion for you to consider in the **Block descriptions** section as follows. 

Original: 

> This description should explain as clearly as possible what your block does.

Suggestion: 

> This description should explain your block's function clearly.

File: docs/design/block-design.md [hunk 1]
```diff
@@ -4,14 +4,85 @@ The following is a light guide to designing a new block with recommendations and
...
+
+Blocks should have a instructive placeholder state when they’re first inserted. If the block includes a text input, provide placeholder text. If your block holds media, include buttons for uploading files and accessing media libraries, as well as drop-zones for drag-and-drop. 
+
+Do:
+
+![placeholder-do](./placeholder-do.png)
+
+Don't
+
+![placeholder-dont](./placeholder-dont.png)
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]: One thing that isn't clear here is the background effect on the placeholder. If the implication there is "stick closely to the component styles we provide, don't try to be too clever/unique" then that should be clarified as well in the writing.

File: docs/design/block-design.md [hunk 2]
```diff
@@ -4,14 +4,85 @@ The following is a light guide to designing a new block with recommendations and
...
-- When selected, your block may surface additional options like input fields or buttons to configure the block directly, if those are necessary for basic operation.
-- Every block should include a description in the “Block” tab of the Settings sidebar. The description should explain as clearly as possible what your block does. Keep it to a single sentence.
-- The “Block” tab of the Settings Sidebar can contain additional block options and configuration, but keep in mind that a user might dismiss the sidebar and never use it. Do not put critical options there.
-- Check how your block looks, feels, and works on all sorts of devices and screen sizes.
+**Blocks**
+
+Blocks should have a simple label for the Block Library. Keep it as short as possible.
+
+Blocks should have an identifying icon, ideally using a single color. Try to avoid using the same icon used by an existing block. The core block icons are based on [Material Design Icons](https://material.io/tools/icons/). Look to that icon set, or to [Dashicons](https://developer.wordpress.org/resource/dashicons/) for style inspiration.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]: I am fairly sure our CSS won't allow a block icon to have multiple colors so you might not need to mention it. (That said, I could be wrong!)

File: docs/design/block-design.md [hunk 3]
```diff
@@ -4,14 +4,85 @@ The following is a light guide to designing a new block with recommendations and
...
+
+Don't:
+
+![blocks-dont](./blocks-dont.png)
+
+---
+
+**Placeholders**
+
+Blocks should have a instructive placeholder state when they’re first inserted. If the block includes a text input, provide placeholder text. If your block holds media, include buttons for uploading files and accessing media libraries, as well as drop-zones for drag-and-drop.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]: I'd clarify this with "If your block requires a user to configure some options before you can display it, you should provide an instructive placeholder state."

Not sure about the placeholder text suggestion — we should check in with a11y to see what the best practice is there.

File: docs/design/block-design.md [hunk 4]
```diff
@@ -4,14 +4,85 @@ The following is a light guide to designing a new block with recommendations and
...
+
+Don't:
+
+![block-controls-dont](./block-controls-dont.png)
+
+*In this example, the block does not appear functional to a user who has dismissed the sidebar.*
+
+---
+
+**Block descriptions**
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]: Small thing but I'd consider grouping this up with the Blocks heading up at the top since setting label/icon/descriptions kind of go together.

**Review by chrisvanpatten — CHANGES_REQUESTED** [review]
Small things but overall it's great!

File: docs/design/block-design.md (hunk 3 above)
Comment (chrisvanpatten) [reply docs/design/block-design.md]: cc @tofumatt :)

**Comment by alexislloyd** [comment]
Thanks for putting this together @jasmussen! I wonder if it might be helpful to separate this into two parts. One part of this is the "dos and don'ts", and I think what you have here works quite well. But another piece, which you touch on but maybe needs more elaboration, is a "Guide to block interactions" (or a slightly better title than that! :). The core question that seems to come up in block design is "where should this control go?", so the more we can provide guidance there, the better. 
Here's a first stab at that:

**The primary interface for a block is the content area of the block**
Since the block itself represents what will actually appear on the site, interaction here hews closest to the principle of direct manipulation and will be most intuitive to the user. This should be thought of as the primary interface for adding and manipulating content and adjusting how it is displayed. There are two ways of interacting here:

1. The placeholder content in the content area of the block can be thought of as a guide or interface for users to follow a set of instructions or “fill in the blanks”. For example, a block that embeds content from a 3rd-party service might contain controls for signing in to that service in the placeholder.

2. After the user has added content, selecting the block can reveal additional controls to adjust or edit that content. For example, a signup block might reveal a control for hiding/showing subscriber count. However, this should be done in minimal ways, so as to avoid dramatically changing the size and display of a block when a user selects it (this could be disorienting or annoying).

**The block toolbar is a secondary place for required options & controls**
Basic block settings won’t always make sense in the context of the placeholder / content UI. As a secondary option, options that are critical to the functionality of a block can live in the block toolbar. The block toolbar is still highly contextual and visible on all screen sizes. One notable constraint with the block toolbar is that it is icon-based UI, so any controls that live in the block toolbar need to be ones that can effectively be communicated via an icon or icon group. 

**The block sidebar should only be used for advanced, tertiary controls.** 
The sidebar is not visible by default on a small / mobile screen, and may also be collapsed in a desktop view. Therefore, it should not be relied on for anything that is necessary for the basic operation of the block. Pick good defaults, make important actions available in the block toolbar, and think of the sidebar as something that only power users may discover. In addition, use sections and headers in the block sidebar if there are more than a handful of options, in order to allow users to easily scan and understand the options available.

**Review by tofumatt — APPROVED** [review]
I had some suggestions/made some tweaks but this looks good to me. I really like the Do/Don't sections; they help me a lot as a developer.

This is better than what we had, so I say 🚢

File: docs/design/block-design.md (hunk 2 above)
Comment (jasmussen) [reply docs/design/block-design.md]: Because they are SVG, they can have multiple colors, and I have actually seen a plugin in the wild using crazy gradient icons. It's looking sort of cool, but is probably bad practice overall.

**Review by davewhitley — CHANGES_REQUESTED** [review]
Just a typo.

**Comment by melchoyce** [comment]
I know this was @drw158's suggestion, but I find the bottom borders hard to follow — my eyes are finding it super hard to scan. What about left borders instead?

image

**Comment by chrisvanpatten** [comment]
I kind of hesitated to post this because we could get deep into bike shedding territory but while we're on the subject of the green/red indicators, one thing to note is that for those of us ~cursed with~ experiencing colour blindness, colours alone aren't necessarily great… on screens especially, where different lighting/backlighting conditions can impact how a colour renders.

Pairing the color bar with a `✔` or `✘`  could go a long way toward making this clearer.

**Comment by drw158** [comment]
@melchoyce I think maybe it's hard to scan because of the different image widths, and the image widths are different from the text widths. Hopefully this won't be a problem in the handbook, just Joen's markdown preview. I wish we had more control over the text color (for the do/don't text), and multi-column layouts.

@chrisvanpatten Do you think the "Do" and "Don't" text alleviates that issue? I could see that being a problem if the do/don't text was missing. 

I wouldn't be against a little check or cross emoji next to the do/don't text though. 

Just wondering, are the red/green WordPress colors optimized for color blindness?

**Comment by chrisvanpatten** [comment]
@drw158 One of my concerns/personal frustrations with only do/don't text below is that it makes scanning a little harder because you have to do a sort of "initial mental parse" to understand whether the text is associated with the image above or below.

I really like the bar on the side, and if that is paired with a visual indicator I think it really aids with scannability.

File: docs/design/block-design.md [hunk 5]
```diff
@@ -1,17 +1,88 @@
...
+
+If your block requires a user to configure some options before you can display it, you should provide an instructive placeholder state.
+
+![A screenshot of the Gallery block's placeholder](./placeholder-do.png)
+**Do:**
+Provide an instructive placeholder state.
+
+![An example Gallery block placeholder but with intense, distracting colors and no instructions](./placeholder-dont.png)
+**Don't:**
+Avoid branding and relying on the title along to convey instructions.
```
Comment (drw158) [inline docs/design/block-design.md]:
```suggestion
Avoid branding and relying on the title alone to convey instructions.
```

**Comment by melchoyce** [comment]
> I think maybe it's hard to scan because of the different image widths, and the image widths are different from the text widths. Hopefully this won't be a problem in the handbook, just Joen's markdown preview.

👍 

> Just wondering, are the red/green WordPress colors optimized for color blindness?

White text on Accent Red (`#DC3232`) passes. White text on Accent Green (`#46B450`) does not. However, if we hop up one shade to `#389547`, large text passes AA. One more hop up to `#31843F` and it totally passes AA. See https://make.wordpress.org/design/handbook/design-guide/foundations/colors/.

**Comment by jasmussen** [comment]
I'm really gosh darn thrilled to see such activity on this PR!

For now, I've fixed the two typos reported by Dave.

As far as I can tell, the subsequent feedback is this:

- What should be the colors of the "Do/Don't" bars
- Where should they be?
- Should there be an additional icon?

Those are all solid points. But I also feel like it's worth shipping this PR _as is_, and then iterating on those aspects separately. We have separate "Do" and "Don't" text labels to help along, which while not the end-all solution, does alleviate the problem for now. 

Here's the Sketch file so anyone can improve these patterns, but I would really appreciate being able to ship this PR for now, and then encourage additional changes separately.

[dosdonts.zip](https://github.com/WordPress/gutenberg/files/2528711/dosdonts.zip)

What do you think?

**Comment by karmatosed** [comment]
I would add the colored lines are unfortunately not helping me with scanning, images or not. I am ok about shipping as is but let's consider the tab (side) design that @melchoyce suggests as that is far more accessible to everyone. Also, let's consider our reliance on colors that could not be read by everyone.

File: docs/design/block-design.md [hunk 6]
```diff
@@ -1,17 +1,88 @@
 # Block Design
 
-The following is a light guide to designing a new block with recommendations and detailed descriptions of existing blocks to illustrate good practices.
+The following are best practices for designed a new block, with recommendations and detailed descriptions of existing blocks to illustrate our approach to creating blocks.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]:
```suggestion
The following are best practices for designing a new block, with recommendations and detailed descriptions of existing blocks to illustrate our approach to creating blocks.
```

File: docs/design/block-design.md [hunk 7]
```diff
@@ -1,17 +1,88 @@
...
+2. After the user has added content, selecting the block can reveal additional controls to adjust or edit that content. For example, a signup block might reveal a control for hiding/showing subscriber count. However, this should be done in minimal ways, so as to avoid dramatically changing the size and display of a block when a user selects it (this could be disorienting or annoying).
+
+### The block toolbar is a secondary place for required options & controls
+Basic block settings won’t always make sense in the context of the placeholder / content UI. As a secondary option, options that are critical to the functionality of a block can live in the block toolbar. The block toolbar is still highly contextual and visible on all screen sizes. One notable constraint with the block toolbar is that it is icon-based UI, so any controls that live in the block toolbar need to be ones that can effectively be communicated via an icon or icon group.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]:
```suggestion
Basic block settings won’t always make sense in the context of the placeholder/content UI. As a secondary option, options that are critical to the functionality of a block can live in the block toolbar. The block toolbar is still highly contextual and visible on all screen sizes. One notable constraint with the block toolbar is that it is an icon-based UI, so any controls that live in the block toolbar need to be ones that can effectively be communicated via an icon or icon group.
```

File: docs/design/block-design.md [hunk 8]
```diff
@@ -1,17 +1,88 @@
...
+
+### The block sidebar should only be used for advanced, tertiary controls.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]: Small consistency tweak here…

```suggestion
### The block sidebar should only be used for advanced, tertiary controls

```

File: docs/design/block-design.md [hunk 9]
```diff
@@ -1,17 +1,88 @@
...
+The sidebar is not visible by default on a small / mobile screen, and may also be collapsed in a desktop view. Therefore, it should not be relied on for anything that is necessary for the basic operation of the block. Pick good defaults, make important actions available in the block toolbar, and think of the sidebar as something that only power users may discover. In addition, use sections and headers in the block sidebar if there are more than a handful of options, in order to allow users to easily scan and understand the options available.
+
+
+
+## Do's and Don'ts 
+
+### Blocks
+
+A block should have a straightforward, short name so users can easily find it in the Block Library. A block named "YouTube" is easy to find and understand. The same block, named "Embedded Video (YouTube)", would be less clear and harder to find in the Block Library.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]:
```suggestion
A block should have a straightforward, short name so users can easily find it in the Block Library. A block named "YouTube" is easy to find and understand. The same block, named "Embedded Video (YouTube)", would be less clear and harder to find in the Block Library. (You can, however, use _keywords_ to make a block searchable with related words. In this case, "embed" and "video" may be appropriate keywords to improve discoverability, while keeping the block name simple.)
```

File: docs/design/block-design.md [hunk 10]
```diff
@@ -1,17 +1,88 @@
...
+
+Blocks should have an identifying icon, ideally using a single color. Try to avoid using the same icon used by an existing block. The core block icons are based on [Material Design Icons](https://material.io/tools/icons/). Look to that icon set, or to [Dashicons](https://developer.wordpress.org/resource/dashicons/) for style inspiration.
+
+![A screenshot of the Block Library with concise block names](./blocks-do.png)
+**Do:**
+Use consice block names.
```
Comment (chrisvanpatten) [inline docs/design/block-design.md]:
```suggestion
Use concise block names.
```

**Review by chrisvanpatten — APPROVED** [review]
Final small tweaks; just a few small things I picked up.

**Comment by chrisvanpatten** [comment]
screenshot 2018-10-30 at 08 12 54

Quick mock 👆

File: docs/design/block-design.md (hunk 9 above)
Comment (jasmussen) [reply docs/design/block-design.md]: I'm slightly nervous about suggesting keywords because in practice I don't feel they've worked super well, and I'm even more concerned about mis-use, even though we have a 3 keyword limit. For example right now "Image" is also a keyword for "Gallery", and since the slash command doesn't yet have any smartness to it, it means if I just inserted a gallery then next time I type `/image` and just press Enter, I get a gallery.

**Comment by jasmussen** [comment]
Okay I took a big final push:

screenshot 2018-10-30 at 13 28 14

I hope with this one we can ship and do smaller iterations in the future. I would also encourage others to create additional do's and don'ts, or edit the existing ones. Here's a new sketch file:

[updated sketch file.zip](https://github.com/WordPress/gutenberg/files/2529611/updated.sketch.file.zip)

Good to go?

**Comment by chrisvanpatten** [comment]
Love it!

**Comment by melchoyce** [comment]
:shipit:

**Review by davewhitley — APPROVED** [review]
